pkg/validators/retry.go

```diff
@@ -0,0 +1,265 @@
+		d.upperLimit = upper
+	}
+}
+
+// ExponentialBackoffLowerLimit set the lower delay limit ensuring that delays
+// do not fall below a minimum duration.
+func ExponentialBackoffLowerLimit(lower time.Duration) ExponentialBackoffOption {
+	return func(d *ExponentialBackoff) {
+		d.lowerLimit = lower
+	}
+}
+
+// ExponentialBackoffRandomizer sets the DurationRandomizer instance which is
+// responsible for generating randomness when applying jitter.
+func ExponentialBackoffRandomizer(rand DurationRandomizer) ExponentialBackoffOption {
+	return func(d *ExponentialBackoff) {
+		d.rand = rand
+	}
+}
+
+// ExponentialBackoffJitter sets the type of jitter to apply when generating
+// new delay values.
+func ExponentialBackoffJitter(jitter Jitter) ExponentialBackoffOption {
+	return func(d *ExponentialBackoff) {
+		d.jitter = jitter
+	}
+}
+
+// DurationRandomizer abstracts behavior which generates random
+// time.Duration values.
+type DurationRandomizer interface {
+	// RandDurationBefore returns a random time.Duration guaranteed
+	// to be less than or equal to the given maximum.
+	RandDurationBefore(max time.Duration) time.Duration
+}
+
+// NewDefaultDurationRandomizer returns an initalized instance of
+// *DefaultDurationRandomizer which seeds it's randomness from
+// the time as of initialization.
+func NewDefaultDurationRandomizer() *DefaultDurationRandomizer {
+	return &DefaultDurationRandomizer{
+		Rand: rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+// DefaultDurationRandomizer provides random duration values
+// generated from a discrete uniform distribution.
+type DefaultDurationRandomizer struct {
+	*rand.Rand
+}
+
+func (dr *DefaultDurationRandomizer) RandDurationBefore(max time.Duration) time.Duration {
+	return time.Duration(dr.Int63n(max.Nanoseconds()))
+}
+
+// Jitter is an enum type describing jitter algorithms implemented in this package.
+type Jitter int
+
+// See https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter for more info
+const (
+	// JitterNone applies no randomness to exponential backoff values
+	JitterNone Jitter = iota
+	// JitterFull applies total randomness to exponential backoff values
+	JitterFull
+	// JitterEqual applies an equal proportion of randomness and determinism
+	// to exponential backoff values. This appears to be a special case of
+	// a more general class of jitter applications which "tunes" the ratio
+	// of jitter.
+	JitterEqual
+	// JitterDecorrellated is not based in exponentiation, but rather scales with
+	// prior random results ala Markov Chains. The choice of '3' as the constant
+	// multiplier in this algorithm is unexplained, but it likely was chosen to
+	// achieve an average curvature more similiar to 2^n.
+	JitterDecorrellated
+)
+
+func durationMin(d1, d2 time.Duration) time.Duration {
+	if d1 < d2 {
+		return d1
+	}
+
+	return d2
+}
+
+func durationMax(d1, d2 time.Duration) time.Duration {
+	if d1 >= d2 {
+		return d1
+	}
+
+	return d2
+}
```

Comment:
:heart: +1 for exponential delay implementation. Read through quickly, I trust you got a good implementation. Off the top of your mind maybe? Rosetta code? :laughing:

Reply:
Uh, a mix off my own knowledge and that AWS blog I mentioned in the comments :)